src/Planner/PlannerJoinTree.cpp

```diff
@@ -606,6 +607,93 @@ std::unique_ptr<ExpressionStep> createComputeAliasColumnsStep(
     return alias_column_step;
 }
 
+std::unique_ptr<ExpressionStep> createAddingVirtualColumnsStep(
+    const DataStream & current_data_stream,
+    const StoragePtr & storage)
+{
+    ColumnWithTypeAndName column;
+    column.name = "_table";
+    column.type = std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>());
+    column.column = column.type->createColumnConst(0, storage->getStorageID().getTableName());
+
+    auto adding_virtual_column_dag = ActionsDAG::makeAddingColumnActions(std::move(column));
+    auto adding_virtual_column_step = std::make_unique<ExpressionStep>(current_data_stream, std::move(adding_virtual_column_dag));
+    adding_virtual_column_step->setStepDescription("Add universal virtual columns");
+    return adding_virtual_column_step;
+}
+
+bool extractRequiredColumnsFromStorage(
+    const Names & columns_names,
+    const StoragePtr & storage,
+    const StorageSnapshotPtr & storage_snapshot,
+    const QueryProcessingStage::Enum processed_stage,
+    SelectQueryInfo & modified_query_info,
+    ContextPtr context,
+    Names & extracted_column_names)
+{
+    if (std::dynamic_pointer_cast<StorageDistributed>(storage))
+    {
+        bool has_table_virtual_column = false;
+        for (const auto & column_name : columns_names)
+        {
+            if (column_name == "_table" && storage->isVirtualColumn(column_name, storage_snapshot->metadata))
+            {
+                has_table_virtual_column = true;
+                break;
+            }
+        }
+
+        if (!has_table_virtual_column)
+            return false;
+
+        const auto & table_name = storage->getStorageID().table_name;
+        auto get_column_options = GetColumnsOptions(GetColumnsOptions::All).withExtendedObjects();
+        if (storage_snapshot->storage.supportsSubcolumns())
+            get_column_options.withSubcolumns();
+
+        std::unordered_map<std::string, QueryTreeNodePtr> column_name_to_node;
+
+        if (!storage_snapshot->tryGetColumn(get_column_options, "_table"))
+        {
+            auto table_name_node = std::make_shared<ConstantNode>(table_name);
+            auto table_name_alias = std::make_shared<ConstantNode>("__table1._table");
+
+            auto function_node = std::make_shared<FunctionNode>("__actionName");
+            function_node->getArguments().getNodes().push_back(std::move(table_name_node));
+            function_node->getArguments().getNodes().push_back(std::move(table_name_alias));
+            function_node->resolveAsFunction(FunctionFactory::instance().get("__actionName", context));
+
+            column_name_to_node.emplace("_table", function_node);
+        }
+
+        if (!column_name_to_node.empty())
+            replaceColumns(modified_query_info.query_tree,
+                modified_query_info.table_expression,
+                column_name_to_node);
+        return false;
```

Comment:
return true

Reply:
For distributed tables, it's special as the table name needed for projection has been transformed into a constant column. Therefore, there's no requirement to add an AddingTableNameVirtualColumnStep subsequently, which is why the return value here is false.